rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+    - l0_last_compacted: copied from the source checkpoint
+    - compacted: copied from the source checkpoint
+    - checkpoints: contains a single entry with the writer’s checkpoint.
+5. If CAS fails, go back to (1) (TODO: we can skip step 2 - we just need to re-validate that the source checkpoint is still valid because the GC could clean it up)
+
+
+##### SST Path Resolution
+
+The DB may now need to look through multiple paths to find SST files, since it may need to open SST files written by a parent db. To support this, we’ll walk the manifest tree up to the root, and build a map from SST ID to Path at startup time, and maintain this in memory.
+
+##### Deleting a Database
+
+With the addition of clones, we need to be a bit more careful about how we delete a database. Previously we could simply remove it's objects from the Object Store and call it a day. With clones, it's possible that a SlateDB instance is holding a checkpoint that another database relies upon. So it's not safe to just delete that data. We'll instead require that users delete a database by calling the `Db::destroy` method. `Db::destroy` can specify that the delete is either a soft or hard delete by using the `soft` parameter. This way we can support both more conservative deletes that clean up data from the GC process after some grace period, and a simple path for hard-deleting a db.
+
+A soft delete will simply mark the manifest as deleted, and rely on a separate GC process to delete the db. It requires that the GC be running in some other long-lived process from the main writer. The GC process will wait for both some grace period, and for all checkpoints to either expire or be deleted. At that point it will delete the db's SSTs and exit. This is covered in more detail in the GC section. A soft delete will be processed by the writer as follows:
+1. Claim write-ownership of the DB by bumping the epoch and fencing the WAL. The intent here is to fence any active writer. We assume that the writer will detect that it's fenced and exit by the time the GC goes to delete the data.
+2. Delete the writer's checkpoint, if any.
+3. Update the manifest by setting the `destroyed_at_s` field to the current time.
+
+A hard delete will proceed as follows. It is up to the user to ensure that there is no active writer at the time of deletion:
+1. Check that there are no active checkpoints. If there are, return `InvalidDeletion`
+2. Delete all objects under `path`.
+
+### Readers
+
+Readers are created by calling `DbReader::open`. `open` takes an optional checkpoint. If a checkpoint is provided, `DbReader` uses the provided checkpoint and does not try to refresh it. If no checkpoint is provided, `DbReader` establishes it’s own checkpoint against the latest manifest and periodically re-establishes this checkpoint at the interval specified in `manifest_poll_interval`. The checkpoints created by `DbReader` use the lifetime specified in the `checkpoint_lifetime` option.
```

Comment:
It would nice to be able to call `create_checkpoint`, get a checkpoint and then use that in `DbReader::open`.

Reply:
Added another comment in `create_checkpoint` section. I am not sure if it is possible to call it without opening the db in write mode.